src/EnergyPlus/Material.cc

```diff
@@ -2758,4 +2763,216 @@ void GetMaterialData(EnergyPlusData &state, bool &ErrorsFound) // set to true if
+        errorsFound = true;
+        return;
+    }
+
+    // Once the material derived type number is found then load the additional CondFD variable material properties
+    //   Some or all may be zero (default).  They will be checked when calculating node temperatures
+    if (addOnType == "Thermal") {
+        auto &thisAbsAddOn = state.dataMaterial->variableThermalAbsAddOns(i);
+        thisAbsAddOn.Name = alphas(1);
+        thisAbsAddOn.MaterialName = alphas(2);
+        thisAbsAddOn.MaterialIdx = MaterNum;
+        thisAbsAddOn.ControlSignal = controlSignal;
+        thisAbsAddOn.FunctionName = alphas(4);
+        thisAbsAddOn.FunctionIdx = functionIdx;
+        thisAbsAddOn.SurfIdx = surfIdx;
+        thisAbsAddOn.ZoneIdx = zoneIdx;
+        thisAbsAddOn.ScheduleName = alphas(5);
+        thisAbsAddOn.ScheduleIdx = scheduleIdx;
+    } else if (addOnType == "Solar") {
+        auto &thisAbsAddOn = state.dataMaterial->variableSolarAbsAddOns(i);
+        thisAbsAddOn.Name = alphas(1);
+        thisAbsAddOn.MaterialName = alphas(2);
+        thisAbsAddOn.MaterialIdx = MaterNum;
+        thisAbsAddOn.ControlSignal = controlSignal;
+        thisAbsAddOn.FunctionName = alphas(4);
+        thisAbsAddOn.FunctionIdx = functionIdx;
+        thisAbsAddOn.SurfIdx = surfIdx;
+        thisAbsAddOn.ZoneIdx = zoneIdx;
+        thisAbsAddOn.ScheduleName = alphas(5);
+        thisAbsAddOn.ScheduleIdx = scheduleIdx;
+    }
+}
+
+void GetVariableThermalAbsorptanceInput(EnergyPlusData &state, bool &errorsFound)
+{
+    int numVariThermalAbs = state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "MaterialProperty:VariableThermalAbsorptance");
+    if (numVariThermalAbs > 0) { //  Get Variable Thermal Absorptance Material Info
+        state.dataMaterial->TotVariableThermalAbsAddOns = numVariThermalAbs;
+        state.dataMaterial->variableThermalAbsAddOns.allocate(numVariThermalAbs);
+        for (int i = 1; i <= numVariThermalAbs; ++i) {
+            ReadingOneMaterialAddOn(state, i, "Thermal", errorsFound);
+        }
+    }
+}
+
+void GetVariableSolarAbsorptanceInput(EnergyPlusData &state, bool &errorsFound)
+{
+    int numVariSolarAbs = state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "MaterialProperty:VariableSolarAbsorptance");
+    state.dataMaterial->TotVariableSolarAbsAddOns = numVariSolarAbs;
+    state.dataMaterial->variableSolarAbsAddOns.allocate(numVariSolarAbs);
+    if (numVariSolarAbs > 0) { //  Get Variable Thermal Absorptance Material Info
+        for (int i = 1; i <= numVariSolarAbs; ++i) {
+            ReadingOneMaterialAddOn(state, i, "Solar", errorsFound);
+        }
+    }
+}
+
+void GetMaterialAddOnData(EnergyPlusData &state, bool &errorsFound)
+{
+    GetVariableThermalAbsorptanceInput(state, errorsFound);
+    GetVariableSolarAbsorptanceInput(state, errorsFound);
+}
+
+// assume only one matching surface
```

Comment:
This is my biggest concern about the implementation. Where does this assumption come from? Why is it valid? And even if it is, why is the chosen surface the first one that matches?

The answers to these questions aside, the implementation here is "inverted". Materials should not reference Zones or Surfaces, that is bad design. You don't want that circularity either in the code (notice "Surfaces have to be set up before this can be set up" and also the large number of otherwise header files you have to include in `Material.cc`) or at runtime. Even if only one Surface needs to know about these new properties, the API for activating this functionality should be attached to the Surface, not to the Material. 

FWIW I notice that the EMS implementation of this feature has the same problem, i.e., it is attached to the material as opposed to each instance of the material. The Material class should be stateless, because it is presumably shared by many Surface instances, each of which will require its own material state.

Reply:
The actual product is a material but the thermal solar absorption for this dynamic coating material is driven by surface properties or zone properties. So is it better to have this coating be added to surfaces properties rather than material?

Reply:
You have to split it up. The state needs to go with Zone and/or Surface, the non-state stuff that is shared by all Surfaces and Zones (ControlSignal, Function, and Schedule) can stay with the Material class.

The more I look at it, there is a lot of this type of confusion in Material and probably elsewhere in EnergyPlus also.

Reply:
Now I think I'm clear. When I apply the overwrite, it will be at the surface/zone level, rather than the material level.

Reply:
Correct. That way each surface gets its own absorptance based on its own temperature and/or the temperature of the zone it is connected to.

Reply:
And you can apply this to all surfaces that use this material rather than just the first one (which is probably just as important).

Reply:
exactly, with the previous design, the user will have to define a different material for each surface, which is the current EMS way. I got confused and followed the EMS way. I should definitely move the state out of the material and make surface reference material.